File: RAO.sh/Microsoft/Day I/2 Combination Sum III/CombinationSumIII.cpp
class Solution{
public:
    void solve(vector<vector<int>> &ans, vector<int> &temp, int k, int n, int start){
        if(n == 0 && temp.size() == k){
            ans.push_back(temp);
            return;
        }
        if(n < 0 || temp.size() > k){
            return;
        }
        for(int i = start; i <= 9; i++){
            temp.push_back(i);
            solve(ans, temp, k, n - i, i + 1);
            temp.pop_back();
        }
    }

    vector<vector<int>> combinationSum3(int k, int n){
        vector<vector<int>> ans;
        vector<int> temp;

        solve(ans, temp, k, n, 1);

        return ans;
    }
};

File: RAO.sh/Microsoft/Day I/2 Combination Sum III/CombinationSumIII.go
package main

import "fmt"

func combinationSum3(k int, n int) [][]int {
	var ans [][]int
	var temp []int

	var solve func(int, int)
	solve = func(start int, sum int) {
		if len(temp) == k {
			if sum == n {
				ans = append(ans, append([]int{}, temp...))
			}
			return
		}

		for i := start; i <= 9; i++ {
			temp = append(temp, i)
			solve(i+1, sum+i)
			temp = temp[:len(temp)-1]
		}
	}

	solve(1, 0)

	return ans
}

func main() {
	n := 7
	k := 3

	var res [][]int
	res = combinationSum3(k, n)

	fmt.Println(res)
}
